added a Feature Histogram Widget

jo-mueller opened the pull request:
Linked to issue #58 

## New features:

This PR introduces a new widget to create Histograms from feature data in Labels/Points/Shapes/Tracks/Vectors layers. It's strongly based off of the ScatterWidget. I wasn't sure whether `scatter.py` was the right place to put it, but it felt like a better fit than `histogram.py`, which essentially just works with image intensity data rather than tabular data.

## Example code snippet

This code snipped demonstrates what it does:

```python
import numpy as np
import napari
from napari_matplotlib import HistogramWidget, FeaturesHistogramWidget

n_points = 1000
random_points = np.random.random((n_points,3))*10
feature1 = np.random.random(n_points)
feature2 = np.random.normal(size=n_points)

viewer = make_napari_viewer()
viewer.add_points(random_points, properties={'feature1': feature1, 'feature2': feature2}, face_color='feature1', size=1)

widget = FeaturesHistogramWidget(viewer)
viewer.window.add_dock_widget(widget)
widget._set_axis_keys('feature1')
widget._key_selection_widget()
widget._set_axis_keys('feature2')
widget._key_selection_widget()
```

This code is essentially also what I added to the tests (tests are passing locally).

Looking forward to hear your thoughts on this! :)

jo-mueller commented:
PS: @dstansby I didn't see that there was ongoing work on a different PR because it wasn't linked to the issue. Feel free to treat this as stale for while.

dstansby commented on src/napari_matplotlib/scatter.py:
```diff
@@ -222,3 +222,124 @@ def _on_update_layers(self) -> None:
+class FeaturesHistogramWidget(NapariMPLWidget):
+    n_layers_input = Interval(1, 1)
+    # All layers that have a .features attributes
+    input_layer_types = (
+        napari.layers.Labels,
+        napari.layers.Points,
+        napari.layers.Shapes,
+        napari.layers.Tracks,
+        napari.layers.Vectors,
+    )
```
Could you put this definition in a new file (`features.py`?) called something like `FEATURES_LAYER_TYPES`, and then use that in both this new widget and the existing features scatter widget?

dstansby commented on src/napari_matplotlib/scatter.py:
The new class should go in `histogram.py`, not this file

jo-mueller replied:
Fixed in f69761871a301a98d8fac475df6cf5a0bdf56c33

dstansby commented on src/napari_matplotlib/scatter.py:
```diff
@@ -222,3 +222,124 @@ def _on_update_layers(self) -> None:
+
+    def __init__(self, napari_viewer: napari.viewer.Viewer):
+        super().__init__(napari_viewer)
+        self.axes = self.canvas.figure.subplots()
```
There's now a `SingleAxesWidget` you can inherit from to avoid having to set up the axes, and to avoid having to define `clear()` below

jo-mueller replied:
Cool! Makes life much easier, thanks for pointing out.

dstansby commented on src/napari_matplotlib/scatter.py:
```diff
@@ -222,3 +222,124 @@ def _on_update_layers(self) -> None:
+        napari.layers.Vectors,
+    )
+
+    def __init__(self, napari_viewer: napari.viewer.Viewer):
```
This should get an optional `parent` argument like all the other widgets.

dstansby commented on src/napari_matplotlib/scatter.py:
```diff
@@ -222,3 +222,124 @@ def _on_update_layers(self) -> None:
+        self._draw()
+
+    def _get_valid_axis_keys(
+        self, combo_widget: Optional[ComboBox] = None
```
I don't think this combo_widget argument is needed

jo-mueller replied:
Fixed in dcab365f898cb945207395af2f0e84afd9170dff

jo-mueller replied:
Fixed in dcab365f898cb945207395af2f0e84afd9170dff

dstansby commented on src/napari_matplotlib/scatter.py:
```diff
@@ -222,3 +222,124 @@ def _on_update_layers(self) -> None:
+        else:
+            return self.layers[0].features.keys()
+
+    def _get_data(self) -> Tuple[List[np.ndarray], str, str]:
```
```suggestion
    def _get_data(self) -> Tuple[np.ndarray, str]:
```

dstansby commented on src/napari_matplotlib/scatter.py:
```diff
@@ -222,3 +222,124 @@ def _on_update_layers(self) -> None:
+        if len(data) == 0:
+            return
+
+        _, _, _ = self.axes.hist(data, bins=50, edgecolor='white',
```
```suggestion
        self.axes.hist(data, bins=50, edgecolor='white',
```

dstansby commented on src/napari_matplotlib/scatter.py:
```diff
@@ -222,3 +222,124 @@ def _on_update_layers(self) -> None:
+        _, _, _ = self.axes.hist(data, bins=50, edgecolor='white',
+                                 linewidth=0.3)
+
+        # # set ax labels
```
```suggestion
        # set ax labels
```

dstansby reviewed (changes requested):
Looking good - a few major changes to make though. I have time to revise this myself, or I can leave you to do it- let me know either way!

jo-mueller commented:
Hi @dstansby ,

thanks a lot for taking the time to review this! I think I can definitely get those changes done myself. I'll be on vacation in the coming week so PR should come sometime in the second-to-next week.

jo-mueller commented:
Hi @dstansby ,

I added your suggestions from above and replaced the magicgui widgets with pyqt code as has been done in the scatter widget in the meantime.

dstansby commented:
@pre-commit-ci autofix

dstansby commented on src/napari_matplotlib/scatter.py:
```diff
@@ -222,6 +217,13 @@ def on_update_layers(self) -> None:
         """
         Called when the layer selection changes by ``self.update_layers()``.
         """
+        if hasattr(self, "_key_selection_widget"):
+            self._key_selection_widget.reset_choices()
+
+        # reset the axis keys
+        self._x_axis_key = None
+        self._y_axis_key = None
```
This should go in the new class, which can call `super().on_update_layers()` as part of it's call.

jo-mueller replied:
@dstansby I'm not sure I got that right - I basically copied line 220 and 221 from the `on_update_layers` method of the `FeaturesScatterWidget`. Is that what you meant?

26a4057286cd2f234e92da4fbc7eebf7b388f249

jo-mueller replied:
> Sorry, think I got a bit confused here, ignore my original comment 

I think I got confused as well and accidentally modified the wrong file 🙈 - will roll back momentarily

dstansby replied:
😆

dstansby reviewed (changes requested):
Looks good! I've left one requst for a change. In addition, could you add a figure test? Hopefully the other test files help with that, but let me know if you need any pointers.

Otherwise I'm 👍 to merge.

jo-mueller commented:
Hi @dstansby ,

I'm not sure what you mean by figure test? I did add the following to the tests. This adds two points layers to the viewer with two sets of features assigned to each. I check whether the figure in the widget changes upon changing the key for the histogram and whether the plot changes as it should when the selected layer is changed.

```python
    n_points = 1000
    random_points = np.random.random((n_points, 3)) * 10
    feature1 = np.random.random(n_points)
    feature2 = np.random.normal(size=n_points)

    viewer = make_napari_viewer()
    viewer.add_points(
        random_points,
        properties={"feature1": feature1, "feature2": feature2},
        name="points1",
    )
    viewer.add_points(
        random_points,
        properties={"feature1": feature1, "feature2": feature2},
        name="points2",
    )

    widget = FeaturesHistogramWidget(viewer)
    viewer.window.add_dock_widget(widget)

    # Check whether changing the selected key changes the plot
    widget._set_axis_keys("feature1")
    fig1 = deepcopy(widget.figure)

    widget._set_axis_keys("feature2")
    assert_figures_not_equal(widget.figure, fig1)

    # check whether selecting a different layer produces the same plot
    viewer.layers.selection.clear()
    viewer.layers.selection.add(viewer.layers[1])
    assert_figures_equal(widget.figure, fig1)
```

dstansby commented:
For a figure test, see e.g.,

https://github.com/matplotlib/napari-matplotlib/blob/87d34f586ae59e596507914479ce8b431b890d50/src/napari_matplotlib/tests/test_histogram.py#L8-L16

This using the `pytest-mpl` plugin to compare the generated image with an image stored in the repository in this directory: https://github.com/matplotlib/napari-matplotlib/tree/main/src/napari_matplotlib/tests/baseline

jo-mueller commented:
For the figure test: So ` @pytest.mark.mpl_image_compare ` will compare the returned image to *any* image in the `baseline` directory?

dstansby commented:
No, you'll need to generate the new test image - see https://github.com/matplotlib/pytest-mpl#usage for instructions.

jo-mueller commented:
@dstansby 

> No, you'll need to generate the new test image - see [matplotlib/pytest-mpl#usage](https://github.com/matplotlib/pytest-mpl#usage) for instructions.

Ok, I added the baseline image, but for some reason, the generated baseline is not detected by the test...Sorry for the back and forth, I'm not so familiar with these testing schemes 🙈 

Just as a sidenote: Setting everything up for `pytest --mpl` seemed quite complicated to me. It requires a few additional installation steps (`pip install pyqt6 tinycss2`) that did not seem evident to me from the.

dstansby commented:
No worries! If you don't mind, I'd be happy to pick this up and get the test working?

jo-mueller commented:
> No worries! If you don't mind, I'd be happy to pick this up and get the test working?

Not at all! Would be cool though if it were documented somewhere for future contributions 👍

dstansby commented:
Looks like the test image was just in the wrong directory :) I also did a few other minor fixes, will merge this and do a new release of napari-matplotlib if the tests pass. Thanks a lot for the contribution, and sorry for how long it took...

dstansby commented:
Looks like it's just code coverage failing
